### ISSUE TITLE
funkcja generateMultipleModels()

### ISSUE THREAD

**Issue by DominikRafacz** [mini-pw/2019L-WarsztatyBadawcze#133]
W skrypcie *multipleOpenMLModelGenerator.R* zawarta jest tytułowa funkcja. Pozwala ona automatycznie generować wiele modeli. Poniżej opisuję jak należy jej używać.

---

Parametry:
**username** - string z nazwą użytkownika, który będzie umieszczony w plikach json
**modelSettings** - obiekt klasy *ModelsSettings*, określający jakie modele funkcja ma wygenerować (szczegóły niżej)
**dataset_id** - numer określający id datasetu w bazie OpenML; domyślnie *NULL* ale nie polecam tego ustawienia (szczegóły niżej)
**target** - string określający nazwę kolumny targetowej; domyślnie *NULL*, funkcja wybiera target domyślny datasetu, ale z powodu niejednolitego nazewnictwa, może być konieczne ręczne podanie
**halt_on_errors** - zmienna logiczna, określająca, czy generator ma się zatrzymać po błędzie przy generowaniu jednego z modeli, czy ma lecieć dalej i tylko wypisać ostrzeżenie; domyślnie *FALSE*
**auto_search** - domyślnie *FALSE* i nie polecam zmieniać (szczegóły niżej)

---

Kod należy wywołać ze ścieżką ustawioną (np. za pomocą *setwd()*) w folderze, w którym chcemy mieć katalogi *openml_\**. Pierwotne założenie jest takie, że jeżeli przed wywołaniem istnieje jeden folder z nazwą *openml_\**, skrypt automatycznie wybiera ten dataset. Jednak z powodu przyczyn (oraz tego, że w folderze *models* nie wszystkie datasety są nazwane tak, jak trzeba), **auto_search** lepiej pozostawić na *FALSE*, a jednak podać **dataset_id**.

---
Działanie:

Kod najpierw sprawdza, czy plik *dataset.json* istnieje. Jeśli nie, tworzy go na podstawie pobranego zbioru. Następnie, jeden po drugim, tworzy modele wyspecyfikowane w parametrze **modelSettings** z uwzględnieniem wszystkich kombinacji przekazanych parametrów.

Parametr modelSettings to obiekt klasy *ModelsSettings* (przepraszam za niejednolite nazewnictwo). Obiekty tej klasy tworzy się za pomocą następującej składni:

MS("nazwa.modelu", par1 = vec1, par2 = vec2, ... ),

gdzie par1, par2 to nazwy parametrów, jakie można przekazać do modelu, a vec1, vec2 to **wektory** wartości, jakie chcemy nadać tym parametrom. Zaznaczam, że są to wektory, ponieważ powyższa funkcja generuje wszystkie możliwe kombinacje tych parametrów. W szczególnym przypadku wektory te mogą być pojedynczymi wartościami, lub w ogóle można nie podawać żadnych parametrów. Na przykład przekazanie poniższego obiektu

MS("classif.ranger", num.trees = c(200, 500), num.iter = c(300, 1000))

do funkcji generateMultipleModels spowoduje wygenerowanie czterech modeli:
makeLearner("classif.ranger", par.vals = list(num.trees = 200, num.iter = 300), predict.type = "prob")
makeLearner("classif.ranger", par.vals = list(num.trees = 200, num.iter = 1000), predict.type = "prob")
makeLearner("classif.ranger", par.vals = list(num.trees = 500, num.iter = 300), predict.type = "prob")
makeLearner("classif.ranger", par.vals = list(num.trees = 500, num.iter = 1000), predict.type = "prob")

Można przekazać dowolną liczbę parametrów, nie trzeba ograniczać się do dwóch. Na przykład przekazanie:

MS("dziki.model", parametr1 = 1:100, parametr2 = letters, parametr3 = c(TRUE, FALSE))

spowoduje wygenerowanie 100 x 26 x 2 modeli. Apeluję jednak o rozsądek. Nie chodzi nam o generowanie modeli ze wszystkimi możliwymi kombinacjami parametrów, raczej ze wszystkimi sensownymi.

Co więcej, te ustawienia można łączyć używając działania dodawania. Widać to chyba na przykładzie:

generateMultipleModels("DominikRafacz",
                       MS("classif.ranger", num.trees = c(200, 500, 1000), num.random.splits = 1:5) +
                       MS("classif.ada", loss = c("exponential", "logistic")) +
                       MS("classif.binomial", link = c("logit" ,"probit", "cloglog", "cauchit")) +
                       MS("classif.kknn"),
                       dataset_id = 40701)

Funkcja nie nadpisuje istniejących plików, więc jeśli ktoś już dodał dany datset i dany model, nie powinien on zostać naruszony. Zapisywanie natomiast jest leniwe - dany model i wszystkie związane z nim pliki są zapisywane dopiero po jego walidacji krzyżowej wykonanej bez błędów, nie ma więc ryzyka, że zapisze się niedokończony model.

Bawcie się dobrze!

**Comment by DominikRafacz** [comment]
Chciałem jeszcze dodać, że tak 50% creditsów leci do Mateusza Bąkały, który jest autorem sporej części funkcji, które wykorzystuję

**Comment by agosiewska** [comment]
SUPER!

**Comment by MatiFilozof** [comment]
Pozwoliłem sobie wprowadzić drobne zmiany. Jako że oryginalny skrypt ograniczony był wyłącznie do używania OpenML, a sam używam przede wszystkim cs.toronto.edu, to uogólniłem kilka rzeczy w kodzie Dominika. W rezultacie do powyższej listy parametrów należy dodać jeszcze jeden.

----

Nowe parametry:
**site** - string z nazwą portalu (wszystkie wielkie litery konwertowane są do małych), przy czym póki co zaimplementowana jest tylko reakcja na `"openml"` i `"toronto"`, gdzie pierwsza z wartości jest wartością domyślną, aby nie psuć wstecznej kompatybilności.

----

Oprócz tego wprowadziłem nową klasę, `DataSet`, która jest listą przechowującą `$data`, `$name`, `$target` i `$targetType`, przy czym kluczowe są pierwsze dwie wartości, a ostatnia prawdopodobnie nawet nie jest używana (ale mogłaby!). Takie rozwiązania wynika z faktu, że OpenML zwraca obiekty klasy `OMLDataSet` przechowujące sporą ilość danych, a inne źródła prawdopodobnie nie (a przynajmniej nie w przypadku cs.toronto.edu), więc konieczne było zbudowanie własnego obiektu, który miałby wymagane pola.

Jeżeli ktoś chciałby dodać inny serwis, to musi dodać odpowiedni przypadek do funkcji `getDataSet` zwracającej obiekt klasy `DataSet` i teoretycznie tyle powinno wystarczyć do prawidłowego działania (przy czym nie zadziała wtedy automatyczne wyszukiwanie zmiennej `target`).
